### PR TITLE
[SPARK-18751][Core]Fix deadlock when SparkContext.stop is called in Utils.tryOrStopSparkContext

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

When `SparkContext.stop` is called in `Utils.tryOrStopSparkContext` (the following three places), it will cause deadlock because the `stop` method needs to wait for the thread running `stop` to exit.

- ContextCleaner.keepCleaning
- LiveListenerBus.listenerThread.run
- TaskSchedulerImpl.start

This PR adds `SparkContext.stopInNewThread` and uses it to eliminate the potential deadlock. I also removed my changes in #15775 since they are not necessary now.

## How was this patch tested?

Jenkins

### CODE REVIEW

**Comment by zsxwing** [comment]
cc @rxin

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -1760,25 +1760,24 @@ class SparkContext(config: SparkConf) extends Logging {
...
 
-        override def run(): Unit = {
-          _stop()
-        }
-      }.start()
-    } else {
-      _stop()
-    }
+      override def run(): Unit = {
+        SparkContext.this.stop()
```
Comment (marmbrus) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Will this ever throw an exception?  Should we register an `UncaughtExceptionHandler` or try catch with logging?

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: > Will this ever throw an exception? Should we register an UncaughtExceptionHandler or try catch with logging?

This happens in the driver, so we cannot use `SparkUncaughtExceptionHandler` to catch the error. The error will be sent to the user's UncaughtExceptionHandler if specified or just print to stderr.

**Comment by vanzin** [comment]
LGTM

**Comment by zsxwing** [comment]
Thanks! Merging to master and 2.1.

**Pull request closed without merging**